components/causal_ts/src/tso.rs

```diff
@@ -0,0 +1,349 @@
+                    .with_label_values(&["ok", reason])
+                    .observe(start.saturating_elapsed_secs());
+                Ok(())
+            }
+        }
+    }
+
+    fn calc_new_batch_size(batch_size: u32, used_size: u32, batch_min_size: u32) -> u32 {
+        if used_size > batch_size * 3 / 4 {
+            // Enlarge to double if used more than 3/4.
+            std::cmp::min(batch_size << 1, TSO_BATCH_MAX_SIZE)
+        } else if used_size < batch_size / 4 {
+            // Shrink to half if used less than 1/4.
+            std::cmp::max(batch_size >> 1, batch_min_size)
+        } else {
+            batch_size
+        }
+    }
+
+    async fn init(&self) -> Result<()> {
+        self.renew_tso_batch(true, TSO_BATCH_RENEW_ON_INITIALIZE)
+            .await?;
+
+        let pd_client = self.pd_client.clone();
+        let tso_batch = self.batch.clone();
+        let batch_min_size = self.batch_min_size;
+        let task = move || {
+            let pd_client = pd_client.clone();
+            let tso_batch = tso_batch.clone();
+            async move {
+                let _ = Self::renew_tso_batch_internal(
+                    pd_client,
+                    tso_batch,
+                    batch_min_size,
+                    false,
+                    TSO_BATCH_RENEW_BY_BACKGROUND,
+                )
+                .await;
+            }
+        };
+
+        // Duration::ZERO means never renew automatically. For test purpose ONLY.
+        if self.renew_interval > Duration::ZERO {
+            self.renew_worker
+                .spawn_interval_async_task(self.renew_interval, task);
+        }
+        Ok(())
+    }
+
+    // Get current batch_size, for test purpose.
+    pub fn batch_size(&self) -> u32 {
+        self.batch.read().size
+    }
+}
+
+const GET_TS_MAX_RETRY: u32 = 3;
+
+impl CausalTsProvider for BatchTsoProvider {
+    fn get_ts(&self) -> Result<TimeStamp> {
+        let start = Instant::now();
+        let mut retries = 0;
+        let mut last_batch_size: u32;
+        loop {
+            {
+                let batch = self.batch.read();
+                last_batch_size = batch.size;
+                match batch.pop() {
+                    Some(ts) => {
+                        trace!("BatchTsoProvider::get_ts: {:?}", ts);
+                        TS_PROVIDER_GET_TS_DURATION
+                            .with_label_values(&["ok"])
+                            .observe(start.saturating_elapsed_secs());
+                        return Ok(ts);
+                    }
+                    None => {
+                        warn!("BatchTsoProvider::get_ts, batch used up"; "batch.size" => batch.size, "retries" => retries);
+                    }
+                }
+            }
+
+            if retries >= GET_TS_MAX_RETRY {
+                break;
+            }
+            if let Err(err) = block_on(self.renew_tso_batch(false, TSO_BATCH_RENEW_FOR_USED_UP)) {
```

Comment:
I suggest to use a channel, so that multiple request to renew tso can be batched up.